File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        # fast
        window = {}
        l = 0
        res = 0

        for r in range(len(s)):
            window[s[r]] = window.get(s[r],0) + 1

            if r-l+1-max(window.values())>k:
                window[s[l]] -= 1
                l += 1

            res = max(res,r-l+1)
        
        return res

File: 0424-longest-repeating-character-replacement/0424-longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        unordered_map<char, int> window;
        int l = 0, res = 0, maxFreq = 0;

        for (int r = 0; r < s.size(); ++r) {
            window[s[r]]++;
            maxFreq = max(maxFreq, window[s[r]]);

            while (r - l + 1 - maxFreq > k) {
                window[s[l]]--;
                if (window[s[l]] == 0) {
                    window.erase(s[l]);
                }
                l++;
            }

            res = max(res, r - l + 1);
        }

        return res;
    }
};